File: scripts/test_solutions/076_minimum_window_substring.js
var minWindow = function(s, t) {
    if (t.length === 0) return '';
    const need = new Map();
    for (const c of t) need.set(c, (need.get(c) || 0) + 1);
    let have = 0, required = need.size;
    let left = 0, minLen = Infinity, minStart = 0;
    const window = new Map();
    for (let right = 0; right < s.length; right++) {
        const c = s[right];
        window.set(c, (window.get(c) || 0) + 1);
        if (need.has(c) && window.get(c) === need.get(c)) have++;
        while (have === required) {
            if (right - left + 1 < minLen) { minLen = right - left + 1; minStart = left; }
            const lc = s[left];
            window.set(lc, window.get(lc) - 1);
            if (need.has(lc) && window.get(lc) < need.get(lc)) have--;
            left++;
        }
    }
    return minLen === Infinity ? '' : s.substring(minStart, minStart + minLen);
};

File: scripts/test_solutions/076_minimum_window_substring.py
from collections import Counter

def minWindow(s, t):
    if not t:
        return ''
    need = Counter(t)
    have, required = 0, len(need)
    left, min_len, min_start = 0, float('inf'), 0
    window = {}
    for right, c in enumerate(s):
        window[c] = window.get(c, 0) + 1
        if c in need and window[c] == need[c]:
            have += 1
        while have == required:
            if right - left + 1 < min_len:
                min_len = right - left + 1
                min_start = left
            lc = s[left]
            window[lc] -= 1
            if lc in need and window[lc] < need[lc]:
                have -= 1
            left += 1
    return '' if min_len == float('inf') else s[min_start:min_start + min_len]
